### PR TITLE
Add types for `core.(molecular_orbitals|operations|sites|spectrum|tensor|xcfunc)`

### PR DESCRIPTION
**Pull request by DanielYang59**
## Summary
- Add/improve types for `core`
- Replace hard-coded class call with `type(self)`
- Replace all `[0:x]` with `[:x]`

### Add types for remaining part of `core`, follow up of #3814.
- [x] `core.molecular_orbitals`
- [x] `core.operations`
- [x] `core.sites`
- [x] `core.spectrum`
- [x] `core.tensor` 
- [x] `core.xcfunc` 

### PR too large, following separated to another PR
- [ ] `core.structure` 
- [ ] `core.units`
- [ ] `core.trajectory`
- Use `type(self)` to avoid hard-coded class name in `__eq__` and similar methods

### CODE REVIEW

File: pymatgen/core/operations.py [hunk 1]
```diff
@@ -55,6 +58,29 @@ def __init__(self, affine_transformation_matrix: ArrayLike, tol: float = 0.01) -
         self.affine_matrix = affine_transformation_matrix
         self.tol = tol
 
+    def __eq__(self, other: object) -> bool:
+        if not isinstance(other, SymmOp):
+            return NotImplemented
+        return np.allclose(self.affine_matrix, other.affine_matrix, atol=self.tol)
+
+    def __hash__(self) -> int:
+        return 7  # DEBUG (DanielYang59): magic hash
```
Comment (DanielYang59) [inline pymatgen/core/operations.py:67]: Another magic `__hash__`

**Comment by DanielYang59** [comment]
For some reason, replacing the magic `__hash__` method of `SymmOp` in bf2a953acfdd3f48a4c379b7030b3c897e43c8bf breaks a unit test of `core.structure` (`test_from_prototype`):

https://github.com/materialsproject/pymatgen/blob/51180fc2ef7ae4208ac244d820f05ce17d57f9c2/pymatgen/core/operations.py#L92-L93

There must be something in it that depends on the hash that I'm not aware of, and I totally have no idea why.

**Review by janosh — APPROVED** [review]
thank a lot @DanielYang59! 👍

**Comment by DanielYang59** [comment]
Thanks for reviewing. Do you have any comments on https://github.com/materialsproject/pymatgen/pull/3829#issuecomment-2109700595?

**Comment by janosh** [comment]
not sure how the `SymmOP.__hash__`  comes into play in that test. you could modify this assert to print actual and expected string to see how they differ

```py
assert str(struct) == expected_struct_str, f"{struct=}, {expected_struct_str=}"
```

**Comment by DanielYang59** [comment]
Thanks for the suggestion, this is what I got:
```python
______________________________________ TestStructure.test_from_prototype ______________________________________

self = <core.test_structure.TestStructure testMethod=test_from_prototype>

        def test_from_prototype(self):
            for prototype in ["bcc", "fcc", "hcp", "diamond"]:
                struct = Structure.from_prototype(prototype, ["C"], a=3, c=4)
                assert isinstance(struct, Structure)
    
            with pytest.raises(ValueError, match="Required parameter 'c' not specified as a kwargs"):
                Structure.from_prototype("hcp", ["C"], a=3)
    
            struct = Structure.from_prototype("rocksalt", ["Li", "Cl"], a=2.56)
            expected_struct_str = """Full Formula (Li4 Cl4)
    Reduced Formula: LiCl
    abc   :   2.560000   2.560000   2.560000
    angles:  90.000000  90.000000  90.000000
    pbc   :       True       True       True
    Sites (8)
      #  SP      a    b    c
    ---  ----  ---  ---  ---
      0  Li    0    0    0
      1  Li    0.5  0.5  0
      2  Li    0.5  0    0.5
      3  Li    0    0.5  0.5
      4  Cl    0.5  0    0.5
      5  Cl    0    0.5  0.5
      6  Cl    0.5  0.5  0
      7  Cl    0    0    0"""
>           assert str(struct) == expected_struct_str, f"{struct=}, {expected_struct_str=}"
E           AssertionError: struct=Structure Summary
E           Lattice
E               abc : 2.56 2.56 2.56
E            angles : 90.0 90.0 90.0
E            volume : 16.777216000000003
E                 A : 2.56 0.0 0.0
E                 B : 0.0 2.56 0.0
E                 C : 0.0 0.0 2.56
E               pbc : True True True
E           PeriodicSite: Li (0.0, 0.0, 0.0) [0.0, 0.0, 0.0]
E           PeriodicSite: Li (0.0, 1.28, 1.28) [0.0, 0.5, 0.5]
E           PeriodicSite: Li (1.28, 0.0, 1.28) [0.5, 0.0, 0.5]
E           PeriodicSite: Li (1.28, 1.28, 0.0) [0.5, 0.5, 0.0]
E           PeriodicSite: Cl (0.0, 1.28, 1.28) [0.0, 0.5, 0.5]
E           PeriodicSite: Cl (1.28, 0.0, 1.28) [0.5, 0.0, 0.5]
E           PeriodicSite: Cl (0.0, 0.0, 0.0) [0.0, 0.0, 0.0]
E           PeriodicSite: Cl (1.28, 1.28, 0.0) [0.5, 0.5, 0.0], expected_struct_str='Full Formula (Li4 Cl4)\nReduced Formula: LiCl\nabc   :   2.560000   2.560000   2.560000\nangles:  90.000000  90.000000  90.000000\npbc   :       True       True       True\nSites (8)\n  #  SP      a    b    c\n---  ----  ---  ---  ---\n  0  Li    0    0    0\n  1  Li    0.5  0.5  0\n  2  Li    0.5  0    0.5\n  3  Li    0    0.5  0.5\n  4  Cl    0.5  0    0.5\n  5  Cl    0    0.5  0.5\n  6  Cl    0.5  0.5  0\n  7  Cl    0    0    0'

/Users/yang/developer/pymatgen/tests/core/test_structure.py:1820: AssertionError
```

`hash` itself doesn't show up in the `str`, and I don't quite know how hash of `SymmOp` is related to generation of `Structure`. I would look closer into this later.

**Pull request merged**